spec/address_geocoder_spec.rb

```diff
@@ -23,7 +23,23 @@
   describe "#valid_address?" do
     context "when address can be not recognized" do
       it "returns false" do
-        address_geocoder = AddressGeocoder.new({api_key: ENV['AddressGeocoderApiKey'], country: 'CN', city: 'Not a city'})
+        # when only city
+        address_geocoder = AddressGeocoder.new({api_key: ENV['AddressGeocoderApiKey'], country: 'CN', city: 'Tokyo'})
+        expect(address_geocoder.valid_address?).to eq(false)
+        # when only postal code
+        address_geocoder = AddressGeocoder.new({api_key: ENV['AddressGeocoderApiKey'], country: 'BR', postal_code: 'A6000A'})
+        expect(address_geocoder.valid_address?).to eq(false)
+        # when only state
+        address_geocoder = AddressGeocoder.new({api_key: ENV['AddressGeocoderApiKey'], country: 'JP', state: 'Ohio'})
+        expect(address_geocoder.valid_address?).to eq(false)
+        # when only street
+        address_geocoder = AddressGeocoder.new({api_key: ENV['AddressGeocoderApiKey'], country: 'CH', street: '10, On Lok Mun Street'})
+        expect(address_geocoder.valid_address?).to eq(false)
+        # when city vs postal code
+        address_geocoder = AddressGeocoder.new({api_key: ENV['AddressGeocoderApiKey'], country: 'CN', city: 'Tokyo', postal_code: '102600'})
+        expect(address_geocoder.valid_address?).to eq(false)
+        # when city vs state
+        address_geocoder = AddressGeocoder.new({api_key: ENV['AddressGeocoderApiKey'], country: 'CN', city: 'Tokyo', state: 'Liaoning'})
```

Comment:
@wingleungchoi that test does a good job of highlighting a problem with the current program. Because of the cycling system of validation, only the uppermost values have to be right for the code to mark it as true. It only takes one match to be considered a valid address, so the state/country match was enough to convince the code that it the address is correct. Possible solution would be lowering the base level to also contain city or by providing a level of certainty with the valid address (eg. .valid_address? could return {valid: true/false, certainty: 'low'/'medium'/'high'}).